subsys/bluetooth/controller/ll_sw/ll_filter.c

```diff
@@ -106,3 +145,344 @@ u32_t ll_wl_remove(bt_addr_le_t *addr)
 	return BT_HCI_ERR_INVALID_PARAM;
 }
 
+#if defined(CONFIG_BLUETOOTH_CONTROLLER_PRIVACY)
+
+#define RL_MATCH(i, id_addr_type, id_addr) (rl[i].taken && \
+		    rl[i].peer_id_addr_type == (id_addr_type & 0x1) && \
+		    !memcmp(rl[i].peer_id_addr.val, id_addr, BDADDR_SIZE))
+
+
+int ll_rl_idx_find(u8_t id_addr_type, u8_t *id_addr)
+{
+	int i;
+	for (i = 0; i < CONFIG_BLUETOOTH_CONTROLLER_RL_SIZE; i++) {
+		if (RL_MATCH(i, id_addr_type, id_addr)) {
+			return i;
+		}
+	}
+
+	return -1;
+}
+
+bool ctrl_rl_enabled(void)
+{
+	return rl_enable;
+}
+
+void ll_rl_pdu_adv_update(int idx, struct pdu_adv *pdu)
+{
+	u8_t *adva = pdu->type == PDU_ADV_TYPE_SCAN_RSP ?
+				  &pdu->payload.scan_rsp.addr[0] :
+				  &pdu->payload.adv_ind.addr[0];
+
+	struct ll_adv_set *ll_adv = ll_adv_set_get();
+
+	/* AdvA */
+	if (idx >= 0 && rl[idx].lirk) {
+		LL_ASSERT(rl[idx].rpas_ready);
+		pdu->tx_addr = 1;
+		memcpy(adva, rl[idx].local_rpa.val, BDADDR_SIZE);
+	} else {
+		pdu->tx_addr = ll_adv->own_addr_type & 0x1;
+		ll_addr_get(ll_adv->own_addr_type & 0x1, adva);
+	}
+
+	/* TargetA */
+	if (pdu->type == PDU_ADV_TYPE_DIRECT_IND) {
+		if (idx >= 0 && rl[idx].pirk) {
+			pdu->rx_addr = 1;
+			memcpy(&pdu->payload.direct_ind.tgt_addr[0],
+			       rl[idx].peer_rpa.val, BDADDR_SIZE);
+		} else {
+			pdu->rx_addr = ll_adv->id_addr_type;
+			memcpy(&pdu->payload.direct_ind.tgt_addr[0],
+			       ll_adv->id_addr, BDADDR_SIZE);
+		}
+	}
+}
+
+static void rpa_adv_refresh(void)
+{
+	struct radio_adv_data *radio_adv_data;
+	struct ll_adv_set *ll_adv;
+	struct pdu_adv *prev;
+	struct pdu_adv *pdu;
+	u8_t last;
+	int idx;
+
+	ll_adv = ll_adv_set_get();
+
+	if (ll_adv->own_addr_type < BT_ADDR_LE_PUBLIC_ID) {
+		return;
+	}
+
+	radio_adv_data = radio_adv_data_get();
+	prev = (struct pdu_adv *)&radio_adv_data->data[radio_adv_data->last][0];
+	/* use the last index in double buffer, */
+	if (radio_adv_data->first == radio_adv_data->last) {
+		last = radio_adv_data->last + 1;
+		if (last == DOUBLE_BUFFER_SIZE) {
+			last = 0;
+		}
+	} else {
+		last = radio_adv_data->last;
+	}
+
+	/* update adv pdu fields. */
+	pdu = (struct pdu_adv *)&radio_adv_data->data[last][0];
+	pdu->type = prev->type;
+	pdu->rfu = 0;
+
+	if (IS_ENABLED(CONFIG_BLUETOOTH_CONTROLLER_CHAN_SEL_2)) {
+		pdu->chan_sel = prev->chan_sel;
+	} else {
+		pdu->chan_sel = 0;
+	}
+
+
+	idx = ll_rl_idx_find(ll_adv->id_addr_type, ll_adv->id_addr);
+	ll_rl_pdu_adv_update(idx, pdu);
+
+	memcpy(&pdu->payload.adv_ind.data[0], &prev->payload.adv_ind.data[0],
+	       prev->len - BDADDR_SIZE);
+	pdu->len = prev->len;;
+
+	/* commit the update so controller picks it. */
+	radio_adv_data->last = last;
+}
+
+static void rl_clear(void)
+{
+	for (int i = 0; i < CONFIG_BLUETOOTH_CONTROLLER_RL_SIZE; i++) {
+		rl[i].taken = 0;
+	}
+}
+
+static int rl_check_access(bool check_ar)
```

Comment:
It's done in ll_rl_clear(). This one doesn't need it as it's the utility function